src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/x64/sysv/CallArranger.java

```diff
@@ -521,9 +526,12 @@ protected BindingCalculator(boolean forArguments) {
     // TODO: handle zero length arrays
     // TODO: Handle nested structs (and primitives)
     private static List<ArgumentClassImpl> classifyStructType(GroupLayout type) {
-        ArgumentClassImpl clazz = (ArgumentClassImpl)Utils.getAnnotation(type, ArgumentClassImpl.ABI_CLASS);
-        if (clazz == ArgumentClassImpl.COMPLEX_X87) {
-            return COMPLEX_X87_CLASSES;
+        var optAbiType = type.abiType();
+        if (optAbiType.isPresent()) {
+            ArgumentClassImpl clazz = SysVx64ABI.argumentClassFor(optAbiType.get());
+            if (clazz == ArgumentClassImpl.COMPLEX_X87) {
+                return COMPLEX_X87_CLASSES;
+            }
```

Comment:
Analog as above, this can be simplified with `Optional::map` to:

````java
if(type.abiType().map(SysVx64ABI::argumentClassFor) == ArgumentClassImpl.COMPLEX_X87) {
    return COMPLEX_X87_CLASSES;
}
````